src/signature.c

```diff
@@ -17,205 +17,426 @@
+static char *crl = NULL;
 
-static char *VERIF_FAIL = "Signature verification failed";
-static char *XSTORE_FAIL = "XSTORE creation failed";
+/* This function must be called before trying to sign any file.
+ * It loads string for errors, and ciphers are auto-loaded by OpenSSL now.
+ * If this function fails it may be because the certificate cannot
+ * be validated.
+ *
+ * returns: true if can initialize and validate certificates, otherwise false */
+static bool initialize_signature(void)
+{
+	ERR_load_crypto_strings();
+	ERR_load_PKCS7_strings();
+	if (!get_pubkey()) {
+		goto fail;
+	}
+	if (!validate_certificate()) {
+		goto fail;
+	}
 
-static bool initialized = false;
+	/* Push our trust cert(s) to the stack, which is a set of certificates
+	 * in which to search for the signer's cert. */
+	x509_stack = sk_X509_new_null();
+	if (!x509_stack) {
+		goto fail;
+	}
+	sk_X509_push(x509_stack, cert);
 
-static X509_STORE *x509_store = NULL;
+	return true;
+fail:
+	return false;
+}
 
-bool signature_initialize(const char *ca_cert_filename)
+/* Delete the memory used for string errors as well as memory allocated for
+ * certificates and private keys. */
+static void terminate_signature(void)
 {
-	if (initialized) {
-		return true;
+	//TODO: once implemented, must free chain
+	//TODO: once implemented, must free crl
+	if (store) {
+		X509_STORE_free(store);
+		store = NULL;
 	}
-	OpenSSL_add_all_algorithms();
-	ERR_load_crypto_strings();
-	x509_store = create_store(ca_cert_filename, NULL, NULL);
-	if (x509_store == NULL) {
-		ERR_free_strings(); // undoes ERR_load_crypto_strings
-		EVP_cleanup();      // undoes OpenSSL_add_all_algorithms
-		return false || (IMPL == IMPL_FORGIVE);
+	if (x509_stack) {
+		sk_X509_pop_free(x509_stack, X509_free);
+		x509_stack = NULL;
 	}
-	initialized = true;
-	return true;
+	ERR_free_strings();
+	if (pkey) {
+		EVP_PKEY_free(pkey);
+		pkey = NULL;
+	}
+	EVP_cleanup();
+	if (fp_pubkey) {
+		fclose(fp_pubkey);
+	}
+	if (cert) {
+		cert = NULL;
+	}
+	ERR_remove_thread_state(NULL);
+	CRYPTO_cleanup_all_ex_data();
 }
 
-void signature_terminate(void)
+/* Verifies that the file and the signature exists, and does a signature
+ * check afterwards.
+ *
+ * returns: true if able to validate the signature, false otherwise */
+static bool verify_signature(const char *data_filename, const char *sig_filename)
 {
-	if (initialized) {
-		X509_STORE_free(x509_store); // undocumented...
-		ERR_free_strings();	  // undoes ERR_load_crypto_strings
-		EVP_cleanup();		     // undoes OpenSSL_add_all_algorithms
-		initialized = false;
+	int ret;
+	struct stat st;
+
+	int data_fd;
+	size_t data_len;
+	unsigned char *data = NULL;
+	BIO *data_BIO;
+
+	int sig_fd;
+	size_t sig_len;
+	unsigned char *sig = NULL;
+	BIO *sig_BIO;
+
+	PKCS7 *p7;
+	BIO *verify_BIO;
+
+	if (!initialize_signature()) {
+		terminate_signature();
+		return false;
+	}
+
+	//this function has issues: EVP_add_digest(EVP_sha256());
+	OpenSSL_add_all_digests();
```

Comment:
Does that work on the second run through the function?

Our big issue this afternoon was appearing to be that we must only call EVP_add_digest() once at process start (not at each run through our verifier function) and must only call the clean up functions one on process termination (not at the end of each run through the verifier function).